### PR TITLE
[Delivers #104925358] Hide reason field until an observer has been selected

### PR DESCRIPTION
**Pull request by yozlet**
Broken out from #649. @pkarman has already reviewed & approved the code. @phirefly, OK to merge?

### CODE REVIEW

File: app/assets/javascripts/hidden_until_select.js.coffee [hunk 1]
```diff
@@ -0,0 +1,16 @@
+class HiddenUntilSelect
+  constructor: ($root, $hidden_el) ->
+    @$hidden_el = $hidden_el
+    @$controller = $root.find("##{ $hidden_el.attr('data-hide-until-select') }")
+    @$controller.keyup => @checkHide()
+    @$controller.change => @checkHide()
```
Comment (afeld) [inline app/assets/javascripts/hidden_until_select.js.coffee]: Are both of these needed? If so, maybe worth an inline comment.

File: app/assets/javascripts/hidden_until_select.js.coffee [hunk 2]
```diff
@@ -0,0 +1,16 @@
+class HiddenUntilSelect
+  constructor: ($root, $hidden_el) ->
+    @$hidden_el = $hidden_el
```
Comment (afeld) [inline app/assets/javascripts/hidden_until_select.js.coffee]: Two very minor things:

* CoffeeScript methods can take a property name prefixed with `@`, which automatically does assignment. These two lines could be collapsed to

    ```javascript
    constructor: ($root, @$hidden_el) ->
    ```

* In JS, the convention is to use camelCase names for variables (`$hiddenEl` vs. `$hidden_el`)

File: app/assets/javascripts/hidden_until_select.js.coffee [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    @$controller.keyup => @checkHide()
+    @$controller.change => @checkHide()
+    @checkHide()
+
+  checkHide: ->
+    @$hidden_el.toggle(@$controller.val())
+
+$ ->
+  # @todo - better scope
+  $scope = $(document)
```
Comment (afeld) [inline app/assets/javascripts/hidden_until_select.js.coffee]: Minor: how about `document.body`?

**Comment by cmc333333** [comment]
Quick note: there's already a `Filter` class. It might make sense to just extend that to fit your use case?

File: app/assets/javascripts/hidden_until_select.js.coffee (hunk 2 above)
Comment (afeld) [inline app/assets/javascripts/hidden_until_select.js.coffee]: Also, `@$hidden_el` is a somewhat misleading name for an element that will eventually be shown. Maybe `@$dependent`?

File: spec/features/observer_spec.rb [hunk 4]
```diff
@@ -53,4 +53,28 @@
...
+    proposal = FactoryGirl.create(:proposal)
+    observer = FactoryGirl.create(:user)
+    login_as(proposal.requester)
+
+    visit "/proposals/#{proposal.id}"
+    submit_button = find("#add_subscriber")
+    expect(submit_button).to be_disabled
+    fill_in_selectized('observation_user_email_address', observer.email_address)
+    expect(submit_button).to_not be_disabled
+  end
```
Comment (afeld) [inline spec/features/observer_spec.rb]: Later TODO: do front-end testing after #658 is in.

File: app/assets/javascripts/hidden_until_select.js.coffee [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  $scope.find("[data-hide-until-select]").each (idx, el) ->
+    new HiddenUntilSelect($scope, $(el))
```
Comment (afeld) [inline app/assets/javascripts/hidden_until_select.js.coffee]: Not necessarily a blocker, but this whole class seems like very similar functionality to the existing [Filter](https://github.com/18F/C2/blob/master/app/assets/javascripts/filter.js.coffee)...we should look to combine them.

Comment (afeld) [inline app/assets/javascripts/hidden_until_select.js.coffee]: Jinx @cmc333333.

Comment (yozlet) [inline app/assets/javascripts/hidden_until_select.js.coffee]: So, after looking at Filter and Aidan's upcoming work, and trying to work out the best way to refactor it all to cover this case, I couldn't see how to do it without adding twice as much code. What Filter does is similar but far enough from identical that it requires a load of extra thought to cover this case.

So I'm going to stick with this solution for the moment, and then we can refactor once we get greater overlap in functionality.

**Comment by yozlet** [comment]
I added a comment which Github seems to have eaten, the gist of which was: while Filter has similar functionality, it's far enough from what's needed here that adding the functionality to Filter would require twice as much code as simply adding this class. But if there's more overlap in the future, we should definitely merge the two.

File: app/assets/javascripts/hidden_until_select.js.coffee [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+  constructor: ($root, @$dependent) ->
+    @$controller = $root.find("##{ @$dependent.attr('data-hide-until-select') }")
+    @$controller.change => @checkHide()
+    @checkHide()
+
+  checkHide: ->
+    @$dependent.toggle(@$controller.val())
+
+$ ->
+  # @todo - better scope
```
Comment (jessieay) [inline app/assets/javascripts/hidden_until_select.js.coffee]: remove `todo` ?

Comment (jessieay) [inline app/assets/javascripts/hidden_until_select.js.coffee]: my view on todos: if it's important but not in scope, it should be in Tracker. If it's important but in scope: should be done now. If it's not important: don't do it!

File: spec/support/feature_spec_helper.rb [hunk 7]
```diff
@@ -10,4 +10,11 @@ def login_as(user)
     setup_mock_auth(:myusa, user)
     visit '/auth/myusa'
   end
+
+  # adapted from http://stackoverflow.com/a/25047358
+  def fill_in_selectized(key, *values)
```
Comment (jessieay) [inline spec/support/feature_spec_helper.rb]: let's put this in the `observer_spec` file as a method inside the `describe` block so it's scoped. Easier to see what's going on that way!

File: spec/features/observer_spec.rb (hunk 4 above)
Comment (yozlet) [inline spec/features/observer_spec.rb]: @afeld: Not entirely sure what you mean here. Fancy adding it as a story?

Comment (afeld) [inline spec/features/observer_spec.rb]: [Done](https://www.pivotaltracker.com/story/show/105635126).

File: coffeelint.json [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+  "space_operators": {
+    "level": "ignore"
+  },
+  "spacing_after_comma": {
+    "level": "ignore"
+  },
+  "transform_messes_up_line_numbers": {
+    "level": "warn"
+  }
+}
```
Comment (afeld) [inline coffeelint.json]: This should probably be in a separate pull request, but please hook it up as part of the CircleCI build! Feel free to move that setup (with the addition of this file) to its own story if you like.

Comment (afeld) [inline coffeelint.json]: Yeah, actually, I'm not sure that I agree with all of these settings so if we could have that discussion separately, that would be great :wink:

**Comment by afeld** [comment]
Just the one comment, to move the adding of `coffeelint.json` to its own PR...otherwise :ship:!

**Pull request merged**